ambari-server/src/main/java/org/apache/ambari/server/agent/AgentReportsProcessor.java

```diff
@@ -37,12 +37,15 @@
 public class AgentReportsProcessor {
   private static final Logger LOG = LoggerFactory.getLogger(AgentReportsProcessor.class);
 
-  private ScheduledExecutorService executor;
+  private final int poolSize;
 
-  private ConcurrentLinkedQueue<AgentReport> agentReportsQueue = new ConcurrentLinkedQueue<>();
+  private final List<ExecutorService> executors;
 
   public void addAgentReport(AgentReport agentReport) {
-    agentReportsQueue.add(agentReport);
+    int hash = agentReport.getHostName().hashCode();
+    hash = hash == Integer.MIN_VALUE ? 0 : hash;
+    int executorNumber = Math.abs(hash) % poolSize;
+    executors.get(executorNumber).submit(new AgentReportProcessingTask(agentReport));
```

Comment:
Since the `Future` returned by `submit()` is ignored, would it make sense to call `execute()` instead?